Migrate airflow-databricks guide (#1064)
astronomer/docs: guides/airflow-databricks.md

---

> title: "Orchestrating Databricks Jobs with Airflow"

Reviewer: ```suggestion
title: "Orchestrating Databricks jobs with Airflow"
```

> sidebar_label: "Orchestrating Databricks Jobs with Airflow"
> description: "Orchestrating Databricks Jobs from your Apache Airflow DAGs."
> id: airflow-databricks
> tags: ["Integrations", "DAGs"]

Reviewer: ```suggestion
sidebar_label: "Orchestrating Databricks jobs with Airflow"
description: "Orchestrating Databricks jobs from your Apache Airflow DAGs."
id: airflow-databricks
tags: [Integrations, DAGs]
```

- Get information about runs or jobs
- Cancel, start, or terminate a cluster
- Install and uninstall libraries on a cluster

### Operators

There are currently two operators in the Databricks provider package: 

- The `DatabricksSubmitRunOperator` makes use of the Databricks [Runs Submit API Endpoint](https://docs.databricks.com/dev-tools/api/latest/jobs.html#runs-submit) and submits a new Spark job run to Databricks.
- The `DatabricksRunNowOperator` makes use of the Databricks [Run Now API Endpoint](https://docs.databricks.com/dev-tools/api/latest/jobs.html#run-now) and runs an existing Spark job.

The `DatabricksRunNowOperator` should be used when you have an existing job defined in your Databricks workspace that you want to trigger using Airflow. The `DatabricksSubmitRunOperator` should be used if you want to manage the definition of your Databricks job and its cluster configuration within Airflow. Both operators allow you to run the job on a Databricks General Purpose cluster you've already created or on a separate Job Cluster that is created for the job and terminated upon the job’s completion. 

> 

Reviewer: ```suggestion
```

> Documentation for both operators can be found [here](https://registry.astronomer.io/providers/databricks?type=operators).

Reviewer: ```suggestion
Documentation for both operators can be found on the [Astronomer Registry](https://registry.astronomer.io/providers/databricks?type=operators).
```

## Example - Using Airflow with Databricks

> Below, is an example DAG that makes use of both the `DatabricksSubmitRunOperator` and the `DatabricksRunNowOperator`. Before diving into the DAG itself, you'll review the prerequisites.

Reviewer: ```suggestion
You'll now learn how to write a DAG that makes use of both the `DatabricksSubmitRunOperator` and the `DatabricksRunNowOperator`. Before diving into the DAG itself, you need to set up your environment to run Databricks jobs.
```

### Create a Databricks connection

> In order to use any Databricks hooks or operators, you will first need to create an Airflow connection that will allow Airflow to talk to your Databricks account. In general, Databricks recommends using a Personal Access Token (PAT) to authenticate to the Databricks REST API. For more information on how to generate a PAT for your account, read the [Managing dependencies in data pipelines](https://docs.databricks.com/dev-tools/data-pipelines.html). 

Reviewer: ```suggestion
In order to use any Databricks hooks or operators, you first need to create an Airflow connection that allows Airflow to talk to your Databricks account. In general, Databricks recommends using a personal access token (PAT) to authenticate to the Databricks REST API. For more information on how to generate a PAT for your account, read the [Managing dependencies in data pipelines](https://docs.databricks.com/dev-tools/data-pipelines.html). 
```

For this example, you'll use the PAT authentication method and set up a connection using the Airflow UI. It should look something like this:

![Databricks Connection](/img/guides/databricks_connection.png)

> The Host should be your Databricks workspace URL, and your PAT should be added as a JSON-formatted Extra.

Reviewer: ```suggestion
The **Host** should be your Databricks workspace URL, and your PAT should be added as a JSON block in **Extra**.
```

}

#Define params for Run Now Operator
notebook_params = {
    "Variable":5
}

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(minutes=2)
}

with DAG('databricks_dag',
    start_date=datetime(2021, 1, 1),
    schedule_interval='@daily',
    catchup=False,
    default_args=default_args
    ) as dag:

    opr_submit_run = DatabricksSubmitRunOperator(
        task_id='submit_run',
        databricks_conn_id='databricks',
        new_cluster=new_cluster,
        notebook_task=notebook_task
    )
    opr_run_now = DatabricksRunNowOperator(
        task_id='run_now',
        databricks_conn_id='databricks',
        job_id=5,
        notebook_params=notebook_params
    )

    opr_submit_run >> opr_run_now
```

For both operators you need to provide the `databricks_conn_id` and necessary parameters. 

For the `DatabricksSubmitRunOperator`, you need to provide parameters for the cluster that will be spun up (`new_cluster`). This should include, at a minimum: 

- Spark version
- Number of workers
- Node type ID 

> These can be defined more granularly as needed. For more information on what Spark version runtimes are available, see [REST API](https://docs.databricks.com/dev-tools/api/latest/index.html#runtime-version-strings).

Reviewer: ```suggestion
These can be defined more granularly as needed. For more information on what Spark version runtimes are available, see the [Databricks REST API documentation](https://docs.databricks.com/dev-tools/api/latest/index.html#runtime-version-strings).
```

### Error handling

When using either of these operators, any failures in submitting the job, starting or accessing the cluster, or connecting with the Databricks API will propagate to a failure of the Airflow task and generate an error message in the logs.

If there is a failure in the job itself, like in one of the notebooks in this example, that failure will also propagate to a failure of the Airflow task. In that case, the error message may not be shown in the Airflow logs, but the logs should include a URL link to the Databricks job status which will include errors, print statements, etc. For example, if you set up the notebook in Job ID 5 in the example above to have a bug in it, you get a failure in the task causing the Airflow task log to look something like this:

![Error Log Example](/img/guides/databricks_failure_airflow_log.png)

In the case above, you can click on the URL link to get to the Databricks log in order to debug the issue.

> ## Where to go from here

Reviewer: ```suggestion
## Next steps
```
